# IPython magic function documentation - `%watermark`

I wrote this simple `watermark` IPython magic function to conveniently add date- and time-stamps to my IPython notebooks. Also, I often want to document various system information, e.g., for my [Python benchmarks](https://github.com/rasbt/One-Python-benchmark-per-day) series.


<br>
<br>

## Installation

The watermark line magic can be installed by executing

```bash
pip install watermark
```

Alternatively, you can install the latest development version directly from GitHub via

```bash
pip install -e git+https://github.com/rasbt/watermark#egg=watermark
```

<br>
<br>

## Loading the `%watermark` magic

To load the `watermark` magic, execute the following line in your IPython notebook or current IPython shell

In [1]:
%load_ext watermark

The version of `watermark` itself can be retrieved as follows:

In [2]:
%watermark --watermark

watermark 2.0.1


<br>
<br>

## Usage

In order to display the optional `watermark` arguments, type

In [3]:
%watermark?

Docstring:
::

  %watermark [-a AUTHOR] [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME]
                 [-v] [-p PACKAGES] [-h] [-m] [-g] [-r] [-b] [-w] [-iv]

IPython magic function to print date/time stamps
and various system information.

optional arguments:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone in the ISO 8601 standard with UTC offset
  -z, --timezone        appends the local time zone
  -u, --updated         appends a string "Last updated: "
  -c CUSTOM_TIME, --custom_time CUSTOM_TIME
                        prints a valid strftime() string
  -v, --python          prints Python and IPython version
  -p PACKAGES, --packages PACKAGES
                        prints ve

<pre>
%watermark [-a AUTHOR] [-d] [-n] [-t] [-i] [-z] [-u] [-c CUSTOM_TIME]
                 [-v] [-p PACKAGES] [-h] [-m] [-g] [-w]

IPython magic function to print date/time stamps
and various system information.

optional arguments:
  -a AUTHOR, --author AUTHOR
                        prints author name
  -d, --date            prints current date as YYYY-mm-dd
  -n, --datename        prints date with abbrv. day and month names
  -t, --time            prints current time as HH-MM-SS
  -i, --iso8601         prints the combined date and time including the time
                        zone the ISO 8601 standard with UTC offset
  -z, --timezone        appends the local time zone
  -u, --updated         appends a string "Last updated: "
  -c CUSTOM_TIME, --custom_time CUSTOM_TIME
                        prints a valid strftime() string
  -v, --python          prints Python and IPython version
  -p PACKAGES, --packages PACKAGES
                        prints versions of specified Python modules and
                        packages
  -h, --hostname        prints the host name
  -m, --machine         prints system and machine info
  -g, --githash         prints current Git commit hash
  -w, --watermark       prints the current version of watermark
</pre>

<br>
<br>

## Examples

In [4]:
%watermark

2019-10-04T18:00:34-05:00

CPython 3.7.1
IPython 7.8.0

compiler   : Clang 4.0.1 (tags/RELEASE_401/final)
system     : Darwin
release    : 18.7.0
machine    : x86_64
processor  : i386
CPU cores  : 12
interpreter: 64bit


<br>

In [5]:
%watermark  -t -d

2019-10-04 18:00:34


In [6]:
%watermark  --iso8601

2019-10-04T18:00:34-05:00


<br>

In [7]:
%watermark -u -n -t -z

last updated: Fri Oct 04 2019 18:00:34 CDT


<br>

In [8]:
%watermark -v

CPython 3.7.1
IPython 7.8.0


<br>

In [9]:
%watermark -m

compiler   : Clang 4.0.1 (tags/RELEASE_401/final)
system     : Darwin
release    : 18.7.0
machine    : x86_64
processor  : i386
CPU cores  : 12
interpreter: 64bit


<br>

In [10]:
%watermark -v -m -p numpy,scipy,sklearn -g

CPython 3.7.1
IPython 7.8.0

numpy 1.17.2
scipy 1.3.1
sklearn 0.21.3

compiler   : Clang 4.0.1 (tags/RELEASE_401/final)
system     : Darwin
release    : 18.7.0
machine    : x86_64
processor  : i386
CPU cores  : 12
interpreter: 64bit
Git hash   : 7b29998a4d445f01664683db656c40158a6b5dee


<br>

In [11]:
%watermark -a "John Doe" -d -v -m

John Doe 2019-10-04 

CPython 3.7.1
IPython 7.8.0

compiler   : Clang 4.0.1 (tags/RELEASE_401/final)
system     : Darwin
release    : 18.7.0
machine    : x86_64
processor  : i386
CPU cores  : 12
interpreter: 64bit


<br>

In [12]:
import scipy as sp
from sklearn import metrics
import numpy.linalg as linalg

In [13]:
%watermark --iversions

scipy   1.3.1
numpy   1.17.2
sklearn 0.21.3

